rfcs/0000-driver-registry.md

```diff
@@ -0,0 +1,379 @@
+
+In practice, any unique/opaque identifier could be used, but a UUID is used for familiarity.
+
+When implementing a driver, the driver would be expected to export its UUID as part of the crate. It's likely the MnemOS project would maintain some registry of "known UUIDs", to attempt to avoid collisions. This could be in a text file, or wiki, etc.
+
+### Using `TypeId`
+
+Although all drivers are accessed by a `KChannel<T>`, which is a heap allocated structure, attempting to hold many channels of different `T`s in the same structure wouldn't work.
+
+Instead, the key:value map would need to type-erase the channel, likely just holding some kind of pointer. Theoretically, the UUID *should* be enough to ensure that the correct `T` matches the expected type, however the cost of getting this wrong (undefined behavior) due to a developer mistake is too high.
+
+In order to avoid this, the `get` function ALSO takes a `TypeId`, which is a unique type identifier provided by the compiler. Since all drivers are statically compiled together, they should all see the same `T` have the same `TypeId`.
+
+This WON'T necessarily work for userspace, or if MnemOS ever gets dynamically loaded drivers, which may be compiled with a different version of Rust, or potentially not Rust at all.
+
+For this reason, non-static users of these drivers will need to use the serialization interface instead. Userspace would send a serialized message to the expected UUID. The kernel would then use the associated deserialization function to deserialize the message, and send it to the associated channel
+
+### Uniqueness of drivers
+
+This scheme requires that only one version of any kind of driver can be registered.
+
+This simplifies things, but could be potentially annoying for certain drivers are expected to have multiple instances, like keyboards, or storage devices.
+
+This proposal sort of ignores that problem, and suggests that those drivers shouldn't be discoverable. Instead, some kind of "higher level" driver that hands out instances of those drivers should be registered.
+
+### (Exclusively) using `KChannel<T>`
+
+Although a lot of discussion of using "message passing" as a primary interface has been had so far, this would pretty much cement this design choice.
+
+Although there are other communication primitives available ready, such as the various flavors of the `bbq` based queues, they could be registered and provided through the KChannel as a "side channel". This is how the current (as of 2022-07-18) serial mux interface works:
+
+* The client who wants to open a virtual port gets the `KProducer`
+* The client sends a "register port" message
+* On success, the client gets back a `bbq` channel handle from the driver
+
+### Interfaces: Kernel Only
+
+If a type used in the channel does NOT support serialization/deserialization, it is limited to only being usable in the kernel, where this step is not required.
```

Comment:
it was not immediately clear from this summary that "kernel only" also includes other drivers in kernelspace, can we mention that here?

Reply:
Do you think a docs update is enough? e.g. in the rfc and method docs for `set_konly`, or does that need another name as well?